Virtual functions to handle memory allocation in Executor::wait_for_work

jacquelinekay opened the pull request:
This PR places the `malloc` and `free` calls in `Executor::wait_for_work()` in two new virtual functions, `get_allocated_handles` and `remove_allocated_handles`. It also creates an example executor, `StaticMemoryExecutor`, which overrides the functions to provide statically allocated arrays for the handles.

wjwwood commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -33,6 +33,8 @@ namespace rclcpp
 namespace executor
 {
 
+enum executor_handle_t {subscriber_handle, service_handle, client_handle, guard_cond_handle};
```
You could use the new C++11 enum types if you were so inclined:

http://www.cprogramming.com/c++11/c++11-nullptr-strongly-typed-enum-class.html

wjwwood commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -215,6 +217,26 @@ class Executor
 
+  /* The default implementation of Executor::get_allocated_handles dynamically
+     allocates the requested handle array, and ignores the parent pointer. */
+  virtual void ** get_allocated_handles(executor_handle_t /*handle_type*/,
```
Maybe these functions should just be `borrow_handles` and `return_handles`? Or maybe `get_handles` and `return_handles`? It seems like using the term allocate here may not always be true.

Also, I guess you're commenting out the parameter name to avoid an unused parameter warning?

wjwwood commented on rclcpp/include/rclcpp/executor.hpp:
```diff
@@ -215,6 +217,26 @@ class Executor
+
+  /* The default implementation of Executor::remove_allocated_handles uses
+     dynamic memory deallocation. */
+  virtual void remove_allocated_handles(void ** handle, size_t /*size*/)
```
I think this function should never need to take the size.

wjwwood commented on rclcpp/include/rclcpp/executors/static_memory_executor.hpp:
```diff
@@ -0,0 +1,133 @@
+#include <rclcpp/macros.hpp>
+#include <rclcpp/node.hpp>
+#include <rclcpp/utilities.hpp>
+#include <rclcpp/rate.hpp>
```
nitpick: can you organize your includes using the style described in the Google C++ style guide?

https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Names_and_Order_of_Includes

wjwwood commented on rclcpp/include/rclcpp/executors/static_memory_executor.hpp:
```diff
@@ -0,0 +1,133 @@
+#define MAX_SUBSCRIPTIONS 16
+#define MAX_SERVICES 8
+#define MAX_CLIENTS 16
+#define MAX_GUARD_CONDS 100
```
Is there a pro/con for having defines here rather than having these be a configuration of the static executor?

wjwwood replied:
I guess you can use static arrays if you don't make them configurable.

wjwwood replied:
At the very least we should make them overridable, maybe like this:

```c++
#ifndef MAX_SUBSCRIPTIONS
#define MAX_SUBSCRIPTIONS 16
#endif
```

The downside to the above is that when we build this into a library that might not be possible for the user to override at that point.

Also, since these are macros and therefore have no implicit name spacing, please prefix them with `RCLCPP_`.

wjwwood commented on rclcpp/include/rclcpp/executors/static_memory_executor.hpp:
```diff
@@ -0,0 +1,133 @@
+  void * subscriptions[MAX_SUBSCRIPTIONS];
+  void * services[MAX_SERVICES];
+  void * clients[MAX_CLIENTS];
+  void * guard_conds[MAX_GUARD_CONDS];
```
We usually put a trailing `_` on private member variables, can you update these to have that too?

wjwwood commented on rclcpp/include/rclcpp/executors/static_memory_executor.hpp:
```diff
@@ -0,0 +1,133 @@
+        if (size > MAX_SUBSCRIPTIONS) {
+          std::cout << "subscriber error" << std::endl;
+          throw std::runtime_error("Size of requested handle pointer exceeded" +
+                  " allocated memory for subscribers.");
```
Looks like you're mixing printing of errors and throwing, I think throwing would be best for now, as it's not really a recoverable situation at the moment.

wjwwood commented on rclcpp/include/rclcpp/executors/static_memory_executor.hpp:
```diff
@@ -0,0 +1,133 @@
+    if (handle == NULL) {
+      std::cout << "warning: Null pointer passed to remove_allocated_handles." <<
+        std::endl;
+      return;
```
I also think throwing is more appropriate here.

wjwwood commented:
Other than a few minor comments this looks like the right approach to me. I guess to follow is a set of changes to fix `get_next_executable`'s use of `new`?

dirk-thomas commented:
I am not sure that *subclassing* is the right approach for this. The way memory allocation happens is a *strategy* which is partly orthogonal to other variations of executors. Therefore I think it would be more flexible and appropriate to use a different pattern (namely dependency injection) to *configure* an executor with the memory allocation strategy.

jacquelinekay commented:
There will be a series of future changes to support different memory
allocation strategies in the executor, so finding the correct pattern is
worthwhile. I will try dependency injection for this problem in a different
branch and we can compare the two approaches.
On Jul 11, 2015 10:21 AM, "Dirk Thomas" <notifications@github.com> wrote:

> I am not sure that *subclassing* is the right approach for this. The way
> memory allocation happens is a *strategy* which is partly orthogonal to
> other variations of executors. Therefore I think it would be more flexible
> and appropriate to use a different pattern (namely dependency injection) to
> *configure* an executor with the memory allocation strategy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/rclcpp/pull/55#issuecomment-120646273>.
>

jacquelinekay commented:
I implemented a first pass at a dependency injection solution for this PR in this branch:
https://github.com/ros2/rclcpp/compare/memory_dependency_injection

I created a simple example of how different memory strategies could be selected dynamically in this branch of `ros2/examples`:
https://github.com/ros2/examples/blob/memory_dependency_injection/rclcpp_examples/src/topics/listener_exec.cpp

I also incorporated most of @wjwwood's comments in this new branch.

@dirk-thomas let me know if this approach looks better to you. I agree that the idea is much better; it allows the developer to flexibly swap out the orthogonal components of Executor. If William agrees, I will close this PR and start a new one.

The main question I have is about the granularity for MemoryStrategy interface. `borrow_handles` and `remove_handles` functions are now functions in MemoryStrategy that can be overridden by child classes (I still think subclassing is the right way to go here), but I also added hooks for `rcl_malloc` and `rcl_free`. A naive example of how the user might use these hooks is in `StaticMemoryStrategy`, but I'm still not sure if this is a good idea.

wjwwood commented:
Lgtm, +1 for replacing the inheritance model for the dep injection.

dirk-thomas commented:
The dep. inj. branch definitielty looks better to me.

jacquelinekay commented:
New pull request:

https://github.com/ros2/rclcpp/pull/56

I also added delegation of instantiating an `AnyExecutable` object for `get_next_ready_executable`.
